### PR TITLE
perf(promise,object): O(1) promise settle tables + per-object dynamic-write gate (#6084 items 2 and 6)

### PR DESCRIPTION
**Pull request by proggeramlug**
Closes the two remaining measured defects on #6084: **item 2** (promise settle is O(N²) in the reaction side tables) and **item 6** (one `Object.freeze` disables the dynamic-write fast path process-wide).

Items 1, 3 (drain), 4 and 5 already landed (#6126, #6257, #6267, #6285). What remains open after this PR is listed at the bottom.

---

## Item 2 — promise settle was O(N²) in three side tables

`js_promise_resolve` / `js_promise_reject` unconditionally drain three promise-pointer-keyed side tables — `PROMISE_SETTLE_LISTENERS`, `PROMISE_OVERFLOW_REACTIONS`, `PROMISE_ALL_STATES`. All three were `RefCell<Vec<(usize, T)>>` **scanned end to end on every settle**, so settling N promises parked in a table is O(N²). The per-sweep death hook (`remove_*_for_dead_promise`) had the same shape: O(table) per dead promise.

### Why it wasn't a data-structure swap

These are **GC root side tables**, and two GC mechanisms depend on the `Vec`:

1. **The incremental root scanner resumes by position** — `PromiseRootScanState { index, slot }` stops mid-table when its step budget runs out and resumes at `entries[index]`. A `HashMap` has no stable positional cursor.
2. **Evacuation rewrites the key in place** — the scanners call `visitor.visit_metadata_usize_slot(&mut entry.key)`; when a moving GC relocates a promise the GC *edits a live entry's key*. Inside a `HashMap` that leaves the entry in the bucket for its old hash — silently unfindable.

### What landed

New `promise/keyed_table.rs`: `PromiseKeyedTable<T>`, the 1:N analogue of the `PromiseContextStore` (#6267) split that `PROMISE_CONTEXTS` already uses.

* The dense `Vec<Entry<T>>` is kept **exactly** as the GC traversal/rewrite surface — the scanners' access pattern and the mutation discipline (append + swap-remove) are unchanged from the raw `Vec`.
* An O(1) `key → positions` index (`PtrHashMap`, so one multiply rather than SipHash on pointer keys) is layered on top as **derived** state.
* Every GC path that can invalidate the index — a key rewritten by evacuation (`note_key_rewritten`), or the copied-minor cleanup dropping/rekeying entries (`retain_mut`) — just sets `index_dirty`. **Nothing patches the index from inside a GC path.** The next lookup rebuilds it from `entries` alone, at most once per cycle that actually moved a promise (and such a cycle already walks the whole table).
* Per-key FIFO order is observable (`p.then(a); p.then(b)`), and the dense `Vec` no longer preserves it once `swap_remove` reorders — so each entry carries a monotonic `seq` and `take_all` sorts by it. `seq` is also what makes the index reconstructible from `entries` alone, which is what lets the GC paths get away with a blunt "mark dirty".

### Measured (`--release`, macOS arm64; min of 7 runs)

N pending promises, each with a 2nd `.then` (so each parks an overflow reaction), timing the settle loop:

| N | main | this PR | node |
|---|---|---|---|
| 5,000 | 17 ms | 1 ms | 0 ms |
| 10,000 | 70 ms | 2 ms | 1 ms |
| 20,000 | 278 ms | 6 ms | 1 ms |
| 40,000 | **1117 ms** | **14 ms** | 1 ms |

**80× at 40k.** The shape is the point: 4× the promises costs main **16.0×** the time (70 → 1117 ms, textbook quadratic) and costs this branch 7× (linear + the constant). The `Promise.all` shape (one reaction per input, which lands in the promise's inline slot) is unchanged: 8 ms vs 8 ms at 40k — no regression.

---

## Item 6 — the write fast path was disabled process-wide

Both dynamic-write fast paths (`js_object_set_field_by_name_transition_fast` and the transition-cache branch inside `js_object_set_field_by_name`) gated on the process-global `GLOBAL_DESCRIPTORS_IN_USE` latch, which flips on **any** descriptor install anywhere and never reverts.

**The issue understates this.** It reports "+29% after an unrelated `Object.freeze`". Measured here, the latch is **already set before user code runs** — the runtime itself installs latching descriptors during startup/first use (an Error's `stack` accessor, `arguments` objects, tagged-template `raw`, typed-array props). So on `main` these fast paths are effectively **dead in every program**, and a program that never calls `Object.freeze` or `defineProperty` at all already pays the O(own-key-count) `[[Set]]` walk on every write:

```
1M objects x 3 new props, no descriptor call anywhere in the program:
  main      2845 ms
  this PR   1465 ms     (1.9x)
```

### What landed

New `plain_data_write_may_intercept(addr, class_id, key)` in `descriptor_state.rs`, used by both gates. It is the same predicate `ordinary_set`'s #5054 fast path (`proxy.rs`) already applies per receiver, and the same receiver-level / prototype-level split as the #5654 read-side guard:

* own descriptors are visible per-object in `OBJ_FLAG_HAS_DESCRIPTORS` (set by `note_descriptor_target`, travels with the object on evacuation, clear on every fresh allocation) — now folded into the existing `FROZEN | SEALED | NO_EXTEND` mask, so it costs nothing;
* only *prototype*-level installs can intercept a write to an object whose own flag is clear, so those are checked against the actual chain: `Object.prototype` **per key** via `object_proto_may_intercept_key` (a blanket check is what made wide builds O(n²), #5054), a recorded `setPrototypeOf` target, or the class chain via `class_instance_set_may_intercept`;
* typed arrays and exotic-expando hosts (whose descriptors `note_descriptor_target` cannot record per-object) are excluded outright;
* when no descriptor exists anywhere, it short-circuits on the same single relaxed load the old gate did — zero added cost in that case.

Conservative in every uncertain case: any doubt returns "may intercept" and takes the slow walk.

### Measured (min of 7)

| | main | this PR |
|---|---|---|
| 1M x 3 props, before any freeze | 2848 ms | **1412 ms** (2.0×) |
| same loop, after `Object.freeze({unrelated:1})` | 3026 ms | **1557 ms** (1.9×) |
| after/before ratio | 1.06 | 1.10 |

Note main's ratio is ~1.0 — *because it was already on the slow path before the freeze*. The freeze isn't what costs you; the latch being pre-set is.

---

## Correctness

This is the part that matters: item 6 **re-enables a path that was dead on `main`**, so its semantics had to be nailed down rather than assumed.

* **Every interception source still intercepts**, verified byte-identical against `node` after an unrelated freeze: inherited setter and inherited non-writable on `Object.prototype`; own accessor and own non-writable on the receiver; frozen and sealed receivers; `setPrototypeOf` and `Object.create` protos carrying a setter; class accessors (vtable); `Object.defineProperty` on a class prototype. Plain data writes on the same objects still take the fast path.
* **Descriptors don't leak across objects** — a frozen/non-writable `x` on object A leaves `b.x = …` freely writable.
* **Reaction order** — 6 reactions on one promise (1 inline slot + 5 overflow) replay in registration order; a promise feeding two `Promise.all` calls settles both; rejection paths fire. Byte-identical to node.
* **GC**: `gc/tests/copying/promise_side_tables.rs` covers exactly the risky path and passes — in particular `test_live_promise_side_table_entries_rekeyed_by_copied_minor`, which parks entries, forces a copied minor that **moves** the promise, and asserts the entries are findable under the **new** address and absent under the stale one. That is the `note_key_rewritten` → index-rebuild path, deterministically. Dead-promise drop and the `PromiseCleanup` finalize arm are covered too.
* A 6k-promise table + 800k-object allocation churn between registration and settle (so GCs run while the tables are populated) replays all 12,000 reactions, matching node, and also under `PERRY_GC_FORCE_EVACUATE=1 PERRY_GC_VERIFY_EVACUATION=1 PERRY_GEN_GC_EVACUATE=1`.
* `cargo test -p perry-runtime`: **1271 passed, 0 failed**.
* Gap suite: no new failures (each pre-existing failure A/B'd byte-for-byte against a `main` build).

### One existing test changed, deliberately

`typed_feedback_object_set_fast_hits_learned_dynamic_key_transition` asserted the *old* semantics:

```rust
if crate::object::descriptors_in_use() {
    assert_eq!(site.fallback_calls, 2);   // global latch set => fast path must bail
} else { ... }
```

That branch encodes exactly the behaviour item 6 removes. It now asserts unconditionally that a fresh plain object with no own descriptor, no recorded `setPrototypeOf` target, and no `Object.prototype` entry for that key hits the learned transition regardless of what descriptors exist elsewhere in the process. (This was the only genuine failure; the other 11 in that module were `TYPED_FEEDBACK_TEST_LOCK` poison cascades from it.)

---

## Tests added

* `crates/perry-runtime/src/promise/keyed_table.rs` — unit tests incl. a randomized differential test against the naive `Vec<(usize, T)>` model (20k ops), rekey/drop-rebuild coverage, and a non-quadratic drain-scaling check.
* `crates/perry/tests/issue_6084_promise_settle_scaling.rs` — settle scaling (4× the promises must not cost ~16×) + reaction-order/multi-registration correctness.
* `crates/perry/tests/issue_6084_write_fast_path_per_object.rs` — every interception source after an unrelated freeze, prototype-chain setters, no cross-object descriptor leakage, and a write-loop ratio guard.

The two perf pins are ratio/scaling based, not absolute-ms, so they survive a loaded CI host while still failing the quadratic and slow-path shapes by a wide margin.

---

## Still open on #6084

* **Timers `BinaryHeap` redesign** (item 3's remaining half) — the `Vec::remove(i)` drain was fixed in #6285; the unsorted-`Vec` scheduler redesign is untouched here.
* `PROMISE_ALL_STATES` is now O(1) like the other two, but note it is not on the hot `Promise.all` path today: `Promise.all` attaches one reaction per input, which lands in the promise's inline reaction slot. The quadratic settle cost in practice comes from `PROMISE_OVERFLOW_REACTIONS` / `PROMISE_SETTLE_LISTENERS` (2nd+ reaction on a promise), which is what the benchmark above exercises.

Separately, and **not** touched here: `await Promise.all([...])` over executor-captured resolvers never resolves on `main` (the `.then` form works). Reproducer below — worth its own issue.

```ts
async function main() {
  const resolvers: ((v: number) => void)[] = [];
  const ps = Array.from({ length: 100 }, () => new Promise<number>((res) => { resolvers.push(res); }));
  const all = Promise.all(ps);
  for (let i = 0; i < 100; i++) resolvers[i](i);
  const r = await all;              // never resumes on main
  console.log("awaited", r.length); // node prints; perry does not
}
main();
```

## Summary by CodeRabbit

* **Performance**
  * Kept dynamic property write fast paths enabled appropriately by vetting intercept behavior per receiver/prototype chain rather than using a process-wide latch.
  * Improved scalability when settling many promises/reactions by using promise-pointer-keyed side tables that avoid quadratic behavior.

* **Bug Fixes**
  * Preserved correct interception behavior for setters, non-writable properties, frozen/sealed objects, and prototype-chain edge cases.
  * Maintained FIFO reaction ordering and correct `Promise.all`/overflow reaction behavior across scenarios.

* **Tests**
  * Added new coverage for per-object write behavior and promise settlement scaling.

### CODE REVIEW

**Comment by proggeramlug** [comment]
### CodeRabbit's Θ(M²) relocation finding: confirmed real, fixed

Good catch — this was right, including the part that stung: **the benchmark that shipped with the PR could not see it.** That benchmark is many keys × *one* entry each (the `Promise.all` shape), so every `Slots` list it ever relocates through has length 1 and the scan is trivially O(1).

#### The defect

`take_all(A)` removes A's entries from the highest position down. Each `swap_remove` displaces the table's **last** entry (always a foreign one) into the vacated slot, and that entry's key's slot list has to be repointed at its new home. `Slots::relocate` found the old position by **scanning** the list — O(M) for a key with M parked entries.

So with all of A's M entries sitting ahead of all of B's M entries, draining A is Θ(M²). This is the exact defect the table exists to remove, just moved out of the table and into a slot list.

It is reachable from plain TypeScript: every `.then()` past the inline slot parks in `PROMISE_OVERFLOW_REACTIONS` keyed by the pending promise, so `for(M) a.then(f); for(M) b.then(f); resolveA()` is all it takes.

#### Measured — Rust, `PromiseKeyedTable` directly (release), time to drain A

| M entries on **each** of 2 keys | before | after |
|---|---|---|
| 20 000 | 41.3 ms | **0.2 ms** |
| 80 000 | 638.6 ms | **0.7 ms** |

Before: 4× the entries per key ⇒ **15.5× the time** (quadratic is 16×). After: 3.5× — linear.

#### Measured — end to end, compiled TypeScript (`a.then()×M; b.then()×M; resolveA()`)

| M | before | after | node 24 |
|---|---|---|---|
| 10 000 | 13 ms | **0 ms** | 1 ms |
| 20 000 | 63 ms | **1 ms** | 1 ms |
| 40 000 | 187 ms | **1 ms** | 2 ms |

#### The fix

Each `Entry` now records `slot`: its own offset within its key's slot list — the reverse of the forward index. A displaced entry names the slot to repoint, so relocation is a single indexed store instead of a scan.

`slot` sits on exactly the same footing as `index` itself, which keeps the invariants this PR established intact:

- the dense `Vec` is still the sole GC traversal/rewrite surface — positional scanner cursor and in-place key rewrite are untouched;
- `slot` is *derived* state: maintained only by `push`/`ensure_index`, meaningless while `index_dirty`, rebuilt wholesale from `entries` alone. **No GC path patches it** — and that is now enforced by the compiler rather than by convention, since `slot` (like `seq`) is a private field, so `scanners.rs` physically cannot reach it through the `pub(super) key`/`value` it does mutate;
- the monotonic `seq` still carries per-key FIFO order.

Note the offset is invariant under position changes, which is what makes repeated displacement of the same entry within one drain correct.

#### Coverage

- **New** `draining_a_heavily_populated_key_ahead_of_another_is_not_quadratic` — the two-heavily-populated-keys regression case. It **fails on the old scan** with the 4×-entries/16×-time signature, so it genuinely pins this.
- `assert_invariants` now also checks the reverse index (each entry must record its own offset in its key's slot list), and the randomized differential-vs-naive-model test now asserts as it goes rather than once at the end — a stale `slot` is only observable through a later relocation.
- `cargo test -p perry-runtime`: 1286 passed / 0 failed.
- Both `issue_6084_*` integration suites green (2 + 4 tests), including `settling_many_promises_with_overflow_reactions_is_not_quadratic` and `unrelated_freeze_does_not_slow_the_write_loop`.
- `gc/tests/copying/promise_side_tables.rs::test_live_promise_side_table_entries_rekeyed_by_copied_minor` still green under `PERRY_GC_FORCE_EVACUATE=1 PERRY_GC_VERIFY_EVACUATION=1`, and `PERRY_GC_DIAG=1` confirms the run is **not vacuous** — `[gc-copy-minor] ran copied_objects=1 copied_bytes=72`, i.e. a promise really was evacuated and its side-table key really was rekeyed.

#### Headline numbers still hold

40k settle **15 ms** (was 1117 ms on main). Item 6 is untouched — this commit changes `promise/keyed_table.rs` and nothing else — and its gate (`unrelated_freeze_does_not_slow_the_write_loop`) is green.

Rebased onto current `main` (which had moved 14 commits, including #6323 on `promise/`).

File: crates/perry-runtime/src/promise/keyed_table.rs [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+                }
+            }
+            Slots::Many(positions) => {
+                for position in positions.iter_mut() {
+                    if *position == old {
+                        *position = new;
+                        return;
+                    }
+                }
+            }
```
[the remark this answers is not in the document]
Comment (proggeramlug) [reply crates/perry-runtime/src/promise/keyed_table.rs:141]: Confirmed and fixed in 0ef55c1 — including your point that the shipped benchmark could not see this (one entry per key ⇒ every relocated `Slots` list has length 1).

Reproduced first, two keys with M entries each, time to drain the one that sits first:

* Rust, `PromiseKeyedTable` directly: M=20k **41.3 ms**, M=80k **638.6 ms** — 4× the entries, 15.5× the time.
* Compiled TS (`a.then(f)×M; b.then(f)×M; resolveA()`, reachable because 2nd+ reactions park in `PROMISE_OVERFLOW_REACTIONS`): M=10k/20k/40k → **13 / 63 / 187 ms**.

After storing each entry's offset within its key's slot list (`Entry::slot`) and rewriting that position directly: **0.2 ms / 0.7 ms** (Rust) and **0 / 1 / 1 ms** (TS, vs node's 1/1/2).

`slot` is derived state on the same footing as `index` — maintained only by `push`/`ensure_index`, meaningless while `index_dirty`, rebuilt wholesale from `entries`, never touched by a GC path (it is a private field, so `scanners.rs` cannot reach it even by accident). Regression test `draining_a_heavily_populated_key_ahead_of_another_is_not_quadratic` added; it fails on the old scanning `relocate`. Full numbers: https://github.com/PerryTS/perry/pull/6327#issuecomment-4954778832
